dipy/segment/clustering.py

```diff
@@ -0,0 +1,372 @@
+        return self._richcmp(other, operator.le)
+
+    def __gt__(self, other):
+        return self._richcmp(other, operator.gt)
+
+    def __ge__(self, other):
+        return self._richcmp(other, operator.ge)
+
+    def add_cluster(self, cluster):
+        """ Adds a new cluster to this cluster map.
+
+        Parameters
+        ----------
+        cluster : `Cluster` object
+            Cluster to be added in this cluster map.
+        """
+        self.clusters.append(cluster)
+        cluster.refdata = self.refdata
+
+    def remove_cluster(self, cluster):
+        """ Remove a cluster from this cluster map.
+
+        Parameters
+        ----------
+        cluster : `Cluster` object
+            Cluster to be removed from this cluster map.
+        """
+        self.clusters.remove(cluster)
+
+    def clear(self):
+        """ Remove all clusters from this cluster map. """
+        self.clusters.clear()
+
+
+class ClusterMapCentroid(ClusterMap):
+    """ Provides functionalities to interact with clustering outputs.
+
+    Useful container to create, remove, retrieve and filter clusters.
+    It allows also to retrieve easely the centroid of every clusters.
+    If `refdata` is given, elements will be returned instead of their
+    index when using `ClusterCentroid` objects.
+
+    Parameters
+    ----------
+    refdata : list
+        Actual elements that clustered indices refer to.
+    """
+    @property
+    def centroids(self):
+        return [cluster.centroid for cluster in self.clusters]
+
+
+class Clustering:
+    def cluster(self, data, ordering=None):
+        """ Clusters `data`.
+
+        Subclasses will perform their clustering algorithm here.
+
+        Parameters
+        ----------
+        data : list of N-dimensional array
+            Each array represents a data point.
+        ordering : iterable of indices, optional
+            Specifies the order in which data points will be clustered.
+
+        Returns
+        -------
+        `ClusterMap` object
+            Result of the clustering.
+        """
+        raise NotImplementedError("Subclass has to define method 'cluster(data, ordering)'!")
+
+
+class QuickBundles(Clustering):
+    r""" Clusters streamlines using QuickBundles algorithm.
+
+    Given a list of streamlines, the QuickBundles algorithm sequentially
+    assigns each streamline to its closest bundle in $\mathcal{O}(Nk)$ where
+    $N$ is the number of streamlines and $k$ is the final number of bundles.
+    If for a given streamline its closest bundle is farther than `threshold`,
+    a new bundle is created and the streamline is assigned to it except if the
+    number of bundles has already exceeded `max_nb_clusters`.
+
+    Parameters
+    ----------
+    threshold : float
+        The maximum distance from a bundle for a streamline to be still
+        considered as part of it.
+    metric : str or `Metric` object
+        The distance metric to use when comparing two streamlines.
```

Comment:
Please explain here the default value which is mdf or better MDF with capital letters and say what it stands for.